pinot-core/src/main/java/com/linkedin/pinot/core/segment/index/readers/OnHeapStringDictionary.java

```diff
@@ -15,162 +15,80 @@
  */
 package com.linkedin.pinot.core.segment.index.readers;
 
-import com.linkedin.pinot.common.Utils;
-import com.linkedin.pinot.core.segment.index.ColumnMetadata;
 import com.linkedin.pinot.core.segment.memory.PinotDataBuffer;
-import java.io.IOException;
-import java.nio.charset.Charset;
-import java.util.HashMap;
-import java.util.Map;
-import joptsimple.internal.Strings;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
+import java.util.Arrays;
 
 
 /**
- * Implementation of String immutable dictionary based using on-heap String array and HashMap.
- * This is useful for:
+ * Implementation of String dictionary that cache all values on-heap.
+ * <p>This is useful for String columns that:
  * <ul>
- *   <li> Low cardinality string dictionaries where memory footprint on-heap is acceptably small. </li>
- *   <li> Heavily queried columns: This helps avoid creation of String from byte[], which is expensive
- *        as well as creates garbage. </li>
+ *   <li>Has low cardinality string dictionary where memory footprint on-heap is acceptably small</li>
+ *   <li>Is heavily queried</li>
  * </ul>
- * This is useful for low cardinality dictionaries that can be stored in-heap with small memory footprint.
- * For
+ * <p>This helps avoid creation of String from byte[], which is expensive as well as creates garbage.
  */
 public class OnHeapStringDictionary extends ImmutableDictionaryReader {
-  private static final Logger LOGGER = LoggerFactory.getLogger(OnHeapStringDictionary.class);
-  private static final Charset UTF_8 = Charset.forName("UTF-8");
-
-  private final int _lengthOfMaxEntry;
-  private final char _paddingChar;
-
-  private String[] _idToStringMap;
-  private String[] _idToRawStringMap;
-  private Map<String, Integer> _stringToIdMap;
-
-  /**
-   * Constructor for the class.
-   *
-   * @param dataBuffer Pinot Data buffer for the dictionary
-   * @param metadata Column metadata
-   */
-  public OnHeapStringDictionary(PinotDataBuffer dataBuffer, ColumnMetadata metadata) {
-    super(dataBuffer, metadata.getCardinality(), metadata.getStringColumnMaxLength());
-    _lengthOfMaxEntry = metadata.getStringColumnMaxLength();
-    _paddingChar = metadata.getPaddingCharacter();
-
-    int cardinality = metadata.getCardinality();
-    _idToStringMap = new String[cardinality];
-    _idToRawStringMap = new String[cardinality];
-    _stringToIdMap = new HashMap<>(cardinality);
-
-    for (int id = 0; id < cardinality; id++) {
-      byte[] bytes = dataFileReader.getBytes(id, 0);
-      String value = getStringFromBytes(bytes, true);
-      _idToStringMap[id] = value;
-      _idToRawStringMap[id] = getStringFromBytes(bytes, false);
-      _stringToIdMap.put(value, id);
+  private final byte _paddingByte;
```

Comment:
Isn't this unnecessary? `_paddingByte` is already included in `ImmutableDictionaryReader`.

Reply:
We need this to decide whether to binary search using unpadded or padded strings